### PR TITLE
Informative error msg for unsupported shells on show()

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

Inspired by a helpdesk call, this adds an informative error message for confused users running Jdaviz outside of a Jupyter context (Voila runs jupyter under the hood, so the command line "Desktop App" should work with this too). It points the users to our docs and sample notebooks. PLEASE TEST THIS BOTH IN JUPYTER AS WELL AS THE COMMAND LINE BEFORE APPROVING!! Thank you devs!

I struggled trying to format this to print out pretty; please let me know if you have ideas on how to better format it!

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by pllim — CHANGES_REQUESTED** [review]
This is a little too drastic for me. I do sometimes debug Jdaviz in IPython session and this would disallow it.

**Comment by duytnguyendtn** [comment]
Just to clarify, a few points/questions: Is the drastic and debugging comment two separate thoughts, or one in the same? Because I do have one point below about the debugging side.

Just to highlight one thing about this PR: this only gets triggered when a user tries to call `show()` from IPython or the Python interpreter. You can still import the package and get the app via `viz.app` as standard, and all other commands work as expected. But please let me know @pllim if this still impacts your workflow so that I may be aware of other scenarios I might not be considering!

**Comment by pllim** [comment]
My protest aside, I think this already breaks the CI.

File: jdaviz/core/helpers.py [hunk 1]
```diff
@@ -315,6 +315,32 @@ def show(self, loc="inline", title=None):
...
+                if type(orig_e) is NameError:
+                    orig_e = RuntimeError("Standard Python Interpreter not supported")
+
+                boilerplate_e = \
+                    RuntimeError('\nYou are currently running Jdaviz from an unsupported '
+                                 'shell. Jdaviz is intended to be run within a Jupyter '
+                                 'notebook, or directly from the command line.\n\nTo run '
+                                 'from Jupyter, call <your viz>.show() from a notebook cell.\n'
+                                 'To run from the command line, run '
+                                 'jdaviz --layout=<your viz> outside of Python.\n\n'
```
Comment (kecnry) [inline jdaviz/core/helpers.py:336]: I think this is out-of-date.  Perhaps just suggest `jdaviz --help`?

File: jdaviz/core/helpers.py [hunk 2]
```diff
@@ -315,6 +315,32 @@ def show(self, loc="inline", title=None):
...
+                                 'To learn more, see "Importing Data" from your config docs at: '
+                                 'https://jdaviz.readthedocs.io/en/latest/index.html\n'
+                                 'For some working examples, see our sample notebooks at: '
+                                 'https://jdaviz.readthedocs.io/en/latest/sample_notebooks.html\n\n'
+                                 'Thanks for trying out Jdaviz! :)')
```
Comment (kecnry) [inline jdaviz/core/helpers.py:341]: I'm not sure this part is necessary, but if you want to leave it in, we should probably format to point to the actual version tag instead of always to latest.  And we'll need to make sure if we ever move `sample_notebooks.html`, that we update this too.

**Review by kecnry — COMMENTED** [review]
I'm always in favor of exposing useful error messages!

I think the CI just needs to be updated to expect the new error being raised, it doesn't look like this actually breaks anything.

All these nested try/excepts and raising errors could be difficult to maintain going forward (they're confusing) and make for a mess of a traceback:

```
RuntimeError: Standard Python Interpreter not supported

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/...../helpers.py", line 342, in show
    raise boilerplate_e from orig_e
RuntimeError:
You are currently running Jdaviz from an unsupported shell. Jdaviz is intended to be run within a Jupyter notebook, or directly from the command line.

To run from Jupyter, call <your viz>.show() from a notebook cell.
To run from the command line, run jdaviz --layout=<your viz> outside of Python.

To learn more, see "Importing Data" from your config docs at: https://jdaviz.readthedocs.io/en/latest/index.html
For some working examples, see our sample notebooks at: https://jdaviz.readthedocs.io/en/latest/sample_notebooks.html

Thanks for trying out Jdaviz! :)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/...../helpers.py", line 368, in show
    raise RuntimeError('Error in displaying Jdaviz') from e
RuntimeError: Error in displaying Jdaviz
```

Is there any reason that just a single check at the top isn't sufficient?

**Comment by pllim** [comment]
As long as CI passes and this does not crash in IPython (not notebook), I will withdraw my disapproval:

```python
from jdaviz import Cubeviz
cubeviz = Cubeviz()
cubeviz.app.data_collection
```

File: jdaviz/core/helpers.py (hunk 1 above)
Comment (duytnguyendtn) [reply jdaviz/core/helpers.py:336]: Whoops, force of habit! Should I correct this or suggest `--help`? I wanted to give the user an answer for their problems, but that puts it on us to keep this message up to date; calling `--help` is probably more sustainable, but puts another step upon the user. What do you think?

**Comment by duytnguyendtn** [comment]
FWIW I removed `test_show.py` because the tests are largely not relevant anymore. Since show reacts to different interpreters, it should always fail in the GH actions interpreter (presumably the standard python interpreter), and we still don't have a way to run the notebook and verify the app "looks" right. If anyone has a better idea how to design a better test, please let me know!

**Comment by kecnry** [comment]
The globals check seems to fail now that its in the actual code rather than directly in the notebook (I'm getting the error on `viz.show()` even within Jupyter)... after playing for a bit, I think the best way to do this and still avoid the try/except is by importing `get_display` from `IPython`, with something like:

```
from IPython import get_display
...
if get_ipython().__class__.__name__ != 'ZMQInteractiveShell':
    raise RuntimeError(...)
```

where when running in python, this will resolve to `'NoneType'`.  Alternatively, an `isinstance` check might be cleaner if there is some access to the class for `ZMQInteractiveShell`, but I'm not sure where to find that.

**Comment by duytnguyendtn** [comment]
Lol I think there was a commit I forgot to push up at the beginning 😅. But thanks for the idea @kecnry ! I think that cleans it up very nicely!

**Review by kecnry — APPROVED** [review]
Works on my end!  Last thing I'll suggest is to just raise the error directly instead of assigning it to a variable (mostly just to avoid that showing up in the traceback and adding any confusion to something whose goal is to avoid confusion).

File: CHANGES.rst [hunk 3]
```diff
@@ -66,9 +68,6 @@ Cubeviz
 - Fixed a bug where having Subset breaks coordinates information display
   in image viewers. [#1472]
 
-Other Changes and Additions
```
Comment (pllim) [inline CHANGES.rst:69]: Why is this removed? Is it on purpose?

File: CHANGES.rst [hunk 4]
```diff
@@ -49,6 +49,8 @@ Specviz
 Other Changes and Additions
 ---------------------------
 
+- Add friendly message when trying to load Jdaviz outside of Jupyter [#1481]
```
Comment (pllim) [inline CHANGES.rst:52]: "friendly" is too subjective.

```suggestion
- Added a more informative error message when trying to load Jdaviz outside of Jupyter [#1481]
```

File: jdaviz/core/helpers.py [hunk 5]
```diff
@@ -314,32 +315,43 @@ def show(self, loc="inline", title=None):
...
+                RuntimeError("\nYou are currently running Jdaviz from an unsupported "
+                             "shell. Jdaviz is intended to be run within a Jupyter "
+                             "notebook, or directly from the command line.\n\n"
+                             "To run from Jupyter, call <your viz>.show() from a notebook cell.\n"
+                             "To see how to run from the command line, run: "
+                             "'jdaviz --help' outside of Python.\n\n"
+                             "To learn more, see our documentation at: "
+                             "https://jdaviz.readthedocs.io\n\n"
+                             "Thanks for trying out Jdaviz! :)")
+            raise boilerplate_e
```
Comment (pllim) [inline jdaviz/core/helpers.py:331]: Smiley face sometimes has the opposite effect (reverse psychology?), so let's make the message less personal.
```suggestion
            raise RuntimeError("\nYou are currently running Jdaviz from an unsupported "
                               f"shell ({cur_shell_name}). Jdaviz is intended to be run within a Jupyter "
                               "notebook, or directly from the command line.\n\n"
                               "To run from Jupyter, call <your viz>.show() from a notebook cell.\n"
                               "To see how to run from the command line, run: "
                               "'jdaviz --help' outside of Python.\n\n"
                               "To learn more, see our documentation at: "
                               "https://jdaviz.readthedocs.io")
```

File: jdaviz/core/helpers.py [hunk 6]
```diff
@@ -314,32 +315,43 @@ def show(self, loc="inline", title=None):
         If "sidecar" is requested in the "classic" Jupyter notebook, the app will appear inline,
         as only JupyterLab has a mechanism to have multiple tabs.
         """
-        try:
-            if loc == "inline":
+        # Check if the user is running Jdaviz in the correct environments.
+        # If not, provide a friendly msg to guide them!
+        if get_ipython().__class__.__name__ != 'ZMQInteractiveShell':
```
Comment (pllim) [inline jdaviz/core/helpers.py:320]:
```suggestion
        cur_shell_name = get_ipython().__class__.__name__
        if cur_shell_name != 'ZMQInteractiveShell':
```

File: jdaviz/core/helpers.py [hunk 7]
```diff
@@ -314,32 +315,43 @@ def show(self, loc="inline", title=None):
...
+                             "notebook, or directly from the command line.\n\n"
+                             "To run from Jupyter, call <your viz>.show() from a notebook cell.\n"
+                             "To see how to run from the command line, run: "
+                             "'jdaviz --help' outside of Python.\n\n"
+                             "To learn more, see our documentation at: "
+                             "https://jdaviz.readthedocs.io\n\n"
+                             "Thanks for trying out Jdaviz! :)")
+            raise boilerplate_e
+
+        if loc == "inline":
```
Comment (pllim) [inline jdaviz/core/helpers.py:333]: Aren't these already in `main`? Why I still see this as diff here? Does this PR need a rebase or did I miss something?

https://github.com/spacetelescope/jdaviz/blob/2386e460b35a1ad77c5490e582a143153ab64b03/jdaviz/core/helpers.py#L318

File: jdaviz/core/helpers.py (hunk 5 above)
Comment (pllim) [reply jdaviz/core/helpers.py:331]: I think this is what Kyle suggested above as well (the raising directly part).

File: jdaviz/core/helpers.py (hunk 7 above)
Comment (kecnry) [reply jdaviz/core/helpers.py:333]: I think just because of the indentation change (and removing the outer try/except)? 🤷

Comment (pllim) [reply jdaviz/core/helpers.py:333]: A rebase would be nice if this hasn't been rebased in a while, just to be on the safe side.

Comment (duytnguyendtn) [reply jdaviz/core/helpers.py:333]: On it!

**Comment by duytnguyendtn** [comment]
@pllim yup, definitely squashing this one

File: jdaviz/core/helpers.py (hunk 7 above)
Comment (duytnguyendtn) [reply jdaviz/core/helpers.py:331]: Rebased and this remains. I believe Kyle's right; this is due to the removal of the upper try block

File: CHANGES.rst (hunk 3 above)
Comment (duytnguyendtn) [reply CHANGES.rst:85]: Nope, not sure how that got in there; I'll revert that

**Comment by pllim** [comment]
> Do we want that?

Well, for now, maybe good enough. If we get a help call saying "error message says my shell is None," then at least you know it is from the interpreter. Some info is better than no info.

**Comment by pllim** [comment]
> I committed the suggestions out of order.

For future reference, you can commit all the suggestions in one batch if you do it over at "File changed" section.

**Comment by duytnguyendtn** [comment]
> all the suggestions in one batch

Yea that was my bad. I generally try to group the suggestions together so they make logical sense, rather than commit them all at the same time, but I didn't realize the two suggestions were dependent on one another 😅

**Comment by duytnguyendtn** [comment]
> If we get a help call saying "error message says my shell is None," then at least you know it is from the interpreter. Some info is better than no info.

Fair enough; I'm happy to let this through. It slightly bugs me that error message is suggesting to me you aren't running in a shell when that *technically* isn't true, but oh well

**Comment by pllim** [comment]
I understand your concern but it is also not really our problem that `get_ipython()` returns `None` sometimes.

**Review by pllim — APPROVED** [review]
I see quite an informative message, so LGTM. Thanks!

```
RuntimeError:
You are currently running Jdaviz from an unsupported shell (TerminalInteractiveShell). Jdaviz is intended to be run within a Jupyter notebook, or directly from the command line.

To run from Jupyter, call <your viz>.show() from a notebook cell.
To see how to run from the command line, run: 'jdaviz --help' outside of Python.

To learn more, see our documentation at: https://jdaviz.readthedocs.io
```

**Comment by pllim** [comment]
dev job failure is unrelated.

**Comment by duytnguyendtn** [comment]
Thanks for the comments/reviews!
